### DISCUSSION TITLE
Understanding FastAPI type checking when parameter is used in dependancy

### DISCUSSION THREAD

**Discussion by gsouveton** [fastapi/fastapi#8693] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum
from fastapi import FastAPI, Depends


class ModelName(str, Enum):
    alexnet = "alexnet"
    resnet = "resnet"
    lenet = "lenet"


def testfunc(model_name: ModelName): # <--- Type hinting here changes behaviour
    return model_name+"123"

app = FastAPI()

@app.get("/models/{model_name}")
async def get_model(model_name: ModelName, test = Depends(testfunc)):
    if model_name == ModelName.alexnet:
        return {"model_name": model_name, "test": test, "message": "Deep Learning FTW!"}

    if model_name.value == "lenet":
        return {"model_name": model_name, "test": test, "message": "LeCNN all the images"}

    return {"model_name": model_name, "test": test, "message": "Have some residuals"}
```

### Description

Hello,

I came across a strange behaviour while progamming an app.
I managed to reproduce it starting from this example code in the doc : https://fastapi.tiangolo.com/tutorial/path-params/#predefined-values

I've just added a dependancy that uses the parameter.

```
...
def testfunc(model_name):
    return model_name+"123"
  
...

@app.get("/models/{model_name}")
async def get_model(model_name: ModelName, test = Depends(testfunc)):
...
```

But once done, the type checking no longer worked for the `model_name` parameter.
In the doc, the field was no longer displayed as a select input but a standard text input.
![default-input](https://user-images.githubusercontent.com/72618454/144220998-4d10e4d6-8139-4bea-89bd-27f020feeb0b.png)

I tried a few things to figure out and finally found that adding the type hint in the dependancy function solved the problem, ensuring validation and displaying a select field in the doc. 
![select-input](https://user-images.githubusercontent.com/72618454/144221059-56f5913c-a507-4762-bf0a-2a50e19a3dca.png)

Is there an explanation to this ? Does the dependancy injection override the type checking by "catching" the parameter directly ?
I couldn't find any technical detail about this behaviour and I would like to be sure to understand it so that I'll be sure not misusing it.

Thank you

### Operating System

Linux

### FastAPI Version

0.70.0

### Python Version

3.9.5

**Comment by YuriiMotov** [comment]
FastAPI relies on parameter type hints to build the schema.
If you use one parameter multiple times (as is in the example provided), you should add type hints every time.
If you omit the type hint, FastAPI will interpret it as `Any`.
In this example, FastAPI analyzes request parameters starting from sub-dependency, finds the `model_name` parameter without type annotation and interprets it as `Any`.
I think FastAPI could emit a warning if the same parameter is used with different type annotations.

Simplified code example:

```py
from fastapi import FastAPI, Depends


def testfunc(model_name: str):
    return model_name+"123"

app = FastAPI()

@app.get("/models/{model_name}")
async def get_model(model_name: int, test = Depends(testfunc)):
    pass
```

<img width="533" height="138" alt="image" src="https://github.com/user-attachments/assets/2aa736ab-d47d-42aa-ae21-3893f8744446" />
